File: packages/core/src/llm/local-model-client.ts
import { createLogger } from '../logging/logger.js';
import { TokenBudgetError } from '../errors.js';
import { withRetry } from '../resilience/api-retry.js';
import { parseJSONResponse } from './json-extract.js';
import type { IClaudeClient, ClaudeResponse } from './claude-client.js';

const log = createLogger('LocalModelClient');

export interface LocalModelClientConfig {
  /** OpenAI 호환 API base URL (예: http://localhost:11434/v1, https://api-inference.huggingface.co/models/meta-llama/Llama-3.1-70B-Instruct/v1) */
  baseUrl: string;
  /** 모델 이름 (예: llama3.1, codellama, deepseek-coder, tgi) */
  model: string;
  maxTokens: number;
  temperature?: number;
  /** API 키. 로컬 모델은 불필요, HuggingFace/OpenRouter 등 클라우드 서비스는 필수. */
  apiKey?: string;
  /** Token budget. 0 = unlimited. */
  tokenBudget?: number;
  /** 요청 타임아웃 (ms). 기본 5분. 로컬 모델은 느릴 수 있다. */
  timeoutMs?: number;
}

/** OpenAI Chat Completion 응답 형식 */
interface ChatCompletionResponse {
  id: string;
  choices: Array<{
    index: number;
    message: { role: string; content: string };
    finish_reason: string;
  }>;
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
}

/**
 * OpenAI 호환 API를 사용하는 범용 모델 클라이언트.
 * 로컬 모델과 OpenAI 호환 클라우드 서비스 모두 지원.
 *
 * 설정 예시:
 * - Ollama: baseUrl='http://localhost:11434/v1', model='llama3.1'
 * - LM Studio: baseUrl='http://localhost:1234/v1', model='local-model'
 * - vLLM: baseUrl='http://localhost:8000/v1', model='meta-llama/...'
 * - HuggingFace: baseUrl='https://api-inference.huggingface.co/models/{model}/v1', model='tgi', apiKey='hf_...'
 * - OpenRouter: baseUrl='https://openrouter.ai/api/v1', model='meta-llama/llama-3.1-70b', apiKey='sk-or-...'
 */
/** SSRF 방어: 클라우드 메타데이터 엔드포인트 차단 */
const BLOCKED_HOSTS = new Set([
  '169.254.169.254',       // AWS EC2 metadata
  'metadata.google.internal', // GCP metadata
  '100.100.100.200',       // Alibaba Cloud metadata
  'fd00:ec2::254',         // AWS EC2 metadata (IPv6)
]);

function validateBaseUrl(baseUrl: string): void {
  let parsed: URL;
  try {
    parsed = new URL(baseUrl);
  } catch {
    throw new Error(`Invalid LOCAL_MODEL_BASE_URL: ${baseUrl}`);
  }
  if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
    throw new Error(`Unsupported protocol in LOCAL_MODEL_BASE_URL: ${parsed.protocol} (only http/https allowed)`);
  }
  if (BLOCKED_HOSTS.has(parsed.hostname)) {
    throw new Error(`Blocked SSRF target in LOCAL_MODEL_BASE_URL: ${parsed.hostname}`);
  }
}

export class LocalModelClient implements IClaudeClient {
  private config: LocalModelClientConfig;
  private totalTokensUsed = 0;

  constructor(config: LocalModelClientConfig) {
    validateBaseUrl(config.baseUrl);
    this.config = config;
    log.debug(
      { baseUrl: config.baseUrl, model: config.model },
      'Local model client initialized',
    );
  }

  get tokensUsed(): number {
    return this.totalTokensUsed;
  }

  async chat(systemPrompt: string, userMessage: string): Promise<ClaudeResponse> {
    const budget = this.config.tokenBudget ?? 0;
    if (budget > 0 && this.totalTokensUsed >= budget) {
      throw new TokenBudgetError(this.totalTokensUsed, budget);
    }

    const response = await withRetry(
      () => this.callApi(systemPrompt, userMessage),
      { maxRetries: 2, baseDelayMs: 1000, maxDelayMs: 10_000 },
      'Local Model',
    );

    this.totalTokensUsed += response.usage.inputTokens + response.usage.outputTokens;

    log.info(
      {
        inputTokens: response.usage.inputTokens,
        outputTokens: response.usage.outputTokens,
        totalUsed: this.totalTokensUsed,
      },
      'Local model usage',
    );

    return response;
  }

  async chatJSON<T>(
    systemPrompt: string,
    userMessage: string,
  ): Promise<{ data: T; usage: ClaudeResponse['usage'] }> {
    const response = await this.chat(
      systemPrompt + '\n\nIMPORTANT: Respond with valid JSON only. No markdown, no explanation.',
      userMessage,
    );

    const data = parseJSONResponse<T>(response.content, 'local model');
    return { data, usage: response.usage };
  }

  private async callApi(systemPrompt: string, userMessage: string): Promise<ClaudeResponse> {
    const url = `${this.config.baseUrl.replace(/\/+$/, '')}/chat/completions`;
    const timeoutMs = this.config.timeoutMs ?? 5 * 60 * 1000;

    const body = {
      model: this.config.model,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userMessage },
      ],
      max_tokens: this.config.maxTokens,
      temperature: this.config.temperature ?? 0.2,
      stream: false,
    };

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), timeoutMs);

    try {
      const headers: Record<string, string> = { 'Content-Type': 'application/json' };
      if (this.config.apiKey) {
        headers['Authorization'] = `Bearer ${this.config.apiKey}`;
      }

      const res = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify(body),
        signal: controller.signal,
      });

      if (!res.ok) {
        const errText = await res.text().catch(() => '');
        throw new Error(`Local model API error ${res.status}: ${errText.slice(0, 200)}`);
      }

      const data = await res.json() as ChatCompletionResponse;

      const content = data.choices[0]?.message?.content ?? '';
      const usage = {
        inputTokens: data.usage?.prompt_tokens ?? 0,
        outputTokens: data.usage?.completion_tokens ?? 0,
      };

      return { content, usage };
    } catch (err) {
      if (err instanceof Error && err.name === 'AbortError') {
        throw new Error(`Local model request timed out after ${timeoutMs}ms`, { cause: err });
      }
      throw err;
    } finally {
      clearTimeout(timeout);
    }
  }
}

File: backend/src/core/llm/local_model_client.py
"""OpenAI 호환 로컬 모델 클라이언트 (Ollama, LM Studio, vLLM, OpenRouter 등)."""
from __future__ import annotations

import httpx

from src.core.errors import AuthError, NetworkError, RateLimitError, TokenBudgetError
from src.core.llm.json_extract import parse_json_response
from src.core.logging.logger import get_logger
from src.core.resilience.api_retry import with_retry

log = get_logger("LocalModelClient")

# SSRF 방어: 메타데이터 엔드포인트 차단
_BLOCKED_HOSTS = {
    "169.254.169.254",
    "metadata.google.internal",
    "169.254.170.2",
    "fd00:ec2::254",
}

TIMEOUT_S = 120.0


def _check_ssrf(url: str) -> None:
    from urllib.parse import urlparse
    host = urlparse(url).hostname or ""
    if host in _BLOCKED_HOSTS:
        raise ValueError(f"SSRF protection: blocked host {host!r}")


class LocalModelClient:
    def __init__(
        self,
        base_url: str,
        model: str,
        api_key: str | None = None,
    ) -> None:
        _check_ssrf(base_url)
        self._base_url = base_url.rstrip("/")
        self._model = model
        self._api_key = api_key
        self._client = httpx.AsyncClient(timeout=TIMEOUT_S)
        self._tokens_used = 0

    async def close(self) -> None:
        await self._client.aclose()

    def _headers(self) -> dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"
        return headers

    async def chat(
        self,
        messages: list[dict],
        system: str | None = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        token_budget: int | None = None,
    ) -> tuple[str, int, int]:
        """
        OpenAI-compatible /v1/chat/completions 호출.
        Returns: (response_text, input_tokens, output_tokens)
        """
        if token_budget and self._tokens_used >= token_budget:
            raise TokenBudgetError(self._tokens_used, token_budget)

        if system:
            messages = [{"role": "system", "content": system}, *messages]

        body = {
            "model": self._model,
            "messages": messages,
            "max_tokens": max_tokens,
            "temperature": temperature,
        }

        async def _call():
            resp = await self._client.post(
                f"{self._base_url}/chat/completions",
                headers=self._headers(),
                json=body,
            )
            # 모든 HTTP 에러를 raise_for_status()로 통일 — 외부 except에서 분류
            resp.raise_for_status()
            return resp.json()

        try:
            data = await with_retry(_call, max_retries=3, label="LocalModel")
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 401:
                raise AuthError("LocalModel", cause=e) from e
            if e.response.status_code == 429:
                raise RateLimitError("LocalModel", cause=e) from e
            raise NetworkError(f"LocalModel HTTP {e.response.status_code}", cause=e) from e
        except httpx.ConnectError as e:
            raise NetworkError(f"Connection failed to {self._base_url}", cause=e) from e
        except httpx.TimeoutException as e:
            raise NetworkError("LocalModel request timed out", cause=e) from e

        choice = data["choices"][0]
        text = choice["message"]["content"] or ""
        usage = data.get("usage", {})
        input_tokens = usage.get("prompt_tokens", 0)
        output_tokens = usage.get("completion_tokens", 0)
        self._tokens_used += input_tokens + output_tokens
        return text, input_tokens, output_tokens

    async def chat_json(
        self,
        messages: list[dict],
        system: str | None = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        token_budget: int | None = None,
    ) -> tuple[dict | list, int, int]:
        text, input_tokens, output_tokens = await self.chat(
            messages=messages, system=system, max_tokens=max_tokens, temperature=temperature,
            token_budget=token_budget,
        )
        return parse_json_response(text), input_tokens, output_tokens
